### ISSUE TITLE
investigate event dispatch optimization and expose a switch for enable/disable the log system

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2534]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#2534|
|Type|enhancement|
|Created|2007-03-21 15:08:39Z|

A Twisted UDP echo server on an 2.2GHz Opteron can saturate a 100mbit/sec network, but doing so requires about 80% CPU.  Since in this application there is essentially no application overhead, this is all time spent in Twisted somewhere.  A similar application in C can do the same task using about 10% CPU.

There are probably some things we can do to cut down on this event dispatch overhead.  One particularly obvious area (but not necessarily the only or even the most costly) is the involvement of the logging system in event dispatch.  Removing the callWithLogger usage from select reactor yields approximately 15% reduction in CPU usage for the same task.  There is probably a way to preserve this functionality without the current overhead, such as by pre-constructing context dictionaries, calling logPrefix only once per selectable, avoiding the extra method call of callWithLogger and going straight to callWithContext, avoiding the context copy since the reactor is guaranteed to be at the top of the context stack, etc.

Attachments:

* [logoff.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/7c2/7c269d45e17c15a6defc8e36c2f9a95852bfa188/4b9e795a894ac4e1f84a017951f8df8c53ed0710.patch) (1412 bytes) - added by pniccoli on 2009-10-15 23:09:03Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__2534 2534
type__enhancement enhancement
reporter__exarkun exarkun
priority__low low
milestone__ 
branch__ 
branch_author__ 
status__new new
resolution__None None
component__core core
keywords__log__cpu_usage log, cpu usage
time__1174489719000000 1174489719000000
changetime__1286996348000000 1286996348000000
version__None None
owner__ 
cc__itamarst cc__therve cc__drewp cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph commented|
|-|-|

As [other recent discussion has revealed](http://twistedmatrix.com/trac/ticket/1272#comment:4), the functionality as currently specified is somewhat ad-hoc and inconsistent.  Making it more regular might make it a lot easier to optimize, so feel free to go down that route; it's just about impossible to write application code that actually relies on the behaviors of the logging system (as they're not only unpredictable, but also mostly undocumented and untested).

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve commented|
|-|-|

I've started to look at this. The current implementation is really really weird and hard to read. From what I understand, this complex system is used for only one thing: keep the logPrefix available in the chain of deferred (and calls in threads), in the system key. The other key, isError, is just to give a default and could trivially be removed.

First, I think logPrefix should be a attribute instead of a method. Nothing use the fact it's a method currently. Of course the pain is backward compatibility (but isn't it always the case ?:).

Then, if we agree on my first point, we could find an easier way to do the same thing.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

Deferreds aren't actually supported, nor threads as far as I recall.  twisted.python.context only works for boring, synchronous call stacks.  It's a way of passing the logPrefix from a place where it is known, inside the reactor, to a place where it is needed, inside the logging code, without having to pass it through all of the intermediate stack frames, which probably include application code which can't be trusted, relied on, or required to pass the information through.

**Comment by twisted-trac** [comment]
|<img alt="pniccoli's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|pniccoli removed owner|
|-|-|

I had the same high cpu usage in my program written in twisted, and after profiling it I've noticed that a lot of time was spent in the logging system. I wrote a simple echo client in twisted, and the performances after disabling the logging system increased of around the 15% (as said above). I know that this is not the correct way to modify it, but if you have tested your application and you are sure it's working, maybe you can use this simple trick to obtain some extra performances.
I added a patch file, in your program you just have to add those two lines:

```
from twisted.python import log
log.logSwitch = False
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

Thanks for your input pniccoli.  It's great to get real-world numbers like this.

(As a matter of workflow, I'm removing the review keyword, since you're not really proposing your attached patch for inclusion in Twisted.)

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
